synapse/app/generic_worker.py

```diff
@@ -178,14 +178,14 @@ async def on_POST(
             # Proxy headers from the original request, such as the auth headers
             # (in case the access token is there) and the original IP /
             # User-Agent of the request.
-            headers = {
-                header: request.requestHeaders.getRawHeaders(header, [])
+            headers: Dict[bytes, List[bytes]] = {
+                header: list(request.requestHeaders.getRawHeaders(header, []))
```

Comment:
AFAICS `getRawHeaders` always returns a list, so I think this is an unfortunate pointless copy. It's needed because Twisted's annotations only promise it returns a Sequence.

IMO this is a bug in Twisted's type hints: as I understand it, the rule of thumb is that functions ought to be liberal in what they accept but precise in what they return.

Reply:
I mean, that rule certainly makes sense but it's also kind of fair enough to give minimal guarantees :/.

Reply:
Should we use a `tuple` instead? My understanding is that `list`s usually have extra allocated space so that you can extend them, but presumably we're not interested in growing these and we'll treat them as immutable anyway.

Reply:
> I mean, that rule certainly makes sense but it's also kind of fair enough to give minimal guarantees :/.

I sympathise. I suppose there are different trade-offs to be made at the library and at the application level.

Shout-out to Hyrum's Law while we're at it.

Reply:
> Should we use a `tuple` instead? My understanding is that `list`s usually have extra allocated space so that you can extend them, but presumably we're not interested in growing these and we'll treat them as immutable anyway.

No strong opinions. I doubt it makes a perceivable difference: this isn't a hot path, and in either case the interpreter will have to alloc a new container + underlying storage and bump the refcounts for every element.